File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/javascript/find_shortest_string.js
function findShortestString(arr) {
  let shortest = arr[0];

  for (let i = 0; i < arr.length; i++) {
    if (arr[i].length < shortest.length) {
      shortest = arr[i];
    }
  }

  return shortest;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  console.log("");

  console.log("Expecting: 'Ron'");
  console.log("=>", findShortestString(['ReBecca', 'Groot', 'Ron', 'Arya']));

  console.log("");

  console.log("Expecting: 'Maine'");
  console.log("=>", findShortestString(['Mississippi', 'Maine', 'Montana', 'Maryland']));

  // BENCHMARK HERE
    const startTime = Date.now();
    let shortInput = ["JavaScript", "Ruby", "Python"];
    let longInput = [];

    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

    for (let i = 0; i < 100; i++) {
      let string = '';
      const length = Math.floor(Math.random() * 10) +1;

      for (let j = 0; j < length; j++) {
        const letter = letters.charAt(Math.floor(Math.random() * letters.length));
        string += letter;
      }

      longInput.push(string);
    }

    for (let i = 0; i < 1000; i++) {
      findShortestString(shortInput);
      findShortestString(longInput);
    }

    let endTime = Date.now();
    let runtime = endTime - startTime;

    console.log(`Runtime: ${runtime} millisecond(s)`);
    console.log(`Average runtime: ${runtime / 2000} millisecond(s)`);
}

module.exports = findShortestString;

// Please add your pseudocode to this file
/*
- Iterate through array
- Set first string to variable
- If any string is shorter than the string stored with the variable, overwrite it
- Return the variable's string
*/


// And a written explanation of your solution
/*
Uses a for loop to check each element of the input array and keep track of the shortest string. In order to be counted as such, each checked string must be shorter than the one already saved, meaning if more than one string is equal length to the another in the array, the one earliest listed will be returned.
*/

File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/ruby/find_shortest_string.rb
def find_shortest_string(arr)
  return arr.reduce do |shortest, str|
    str.length < shortest.length ? str : shortest
  end
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  # Don't forget to add your own!
  puts

  puts "Expecting: 'Ron'"
  puts "=>", find_shortest_string(['ReBecca', 'Groot', 'Ron', 'Arya'])

  puts

  puts "Expecting: 'Maine'"
  puts "=>", find_shortest_string(['Mississippi', 'Maine', 'Montana', 'Maryland'])


  # BENCHMARK HERE
  start_time = Time.now
  short_input = ["JavaScript", "Ruby", "Python"]
  long_input = []

  letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'

  100.times do
    string = ''
    length = rand(10) + 1

    i = 0
    while i < length do
      letter = letters[(rand * letters.length).floor]
      string += letter
      i += 1
    end

    long_input.push(string)
  end

  puts long_input

  1000.times do
    find_shortest_string(short_input)
    find_shortest_string(long_input)
  end

  end_time = Time.now

  puts "Average runtime: #{(end_time - start_time) / 2000} millisecond(s)"
end

# Please add your pseudocode to this file
=begin
-Return the result of the built-in `reduce` method, using an accumulator parameter and a parameter representing each element of the array.
=end

# And a written explanation of your solution
Uses the built-in `reduce` method to keep track of the shortest string so far and the current string from the input array. Returns result of `reduce`, which is from a ternary that results in either the shortest string or the current string, whichever is shorter.
